src/rust/engine/fs/src/lib.rs

```diff
@@ -729,29 +900,137 @@ pub trait VFS<E: Send + Sync + 'static>: Clone + Send + Sync + 'static {
         expansion
           .todo
           .drain(..)
-          .map(|path_glob| context.expand_single(path_glob, exclude))
+          .map(|sourced_glob| context.expand_single(sourced_glob, exclude))
           .collect::<Vec<_>>()
       });
-      round.map(move |paths_and_globs| {
+      round.map(move |single_expansion_results| {
         // Collect distinct new PathStats and PathGlobs
-        for (paths, globs) in paths_and_globs.into_iter() {
-          expansion.outputs.extend(paths.into_iter().map(|p| (p, ())));
-          let completed = &mut expansion.completed;
+        for exp in single_expansion_results {
+          let SingleExpansionResult {
+            sourced_glob: GlobWithSource { path_glob, source },
+            path_stats,
+            globs,
+          } = exp;
+
+          expansion.outputs.extend(path_stats.clone());
+
           expansion
-            .todo
-            .extend(globs.into_iter().filter(|pg| completed.insert(pg.clone())));
+            .completed
+            .entry(path_glob.clone())
+            .or_insert_with(|| GlobExpansionCacheEntry {
+              globs: globs.clone(),
+              matched: if path_stats.is_empty() {
+                GlobMatch::DidNotMatchAnyFiles
+              } else {
+                GlobMatch::SuccessfullyMatchedSomeFiles
+              },
+              sources: vec![],
+            })
+            .sources
+            .push(source);
+
+          // TODO(cosmicexplorer): is cloning these so many times as `GlobSource`s something we can
```

Comment:
We may want to follow up on this in a future PR, so would leave this as a TODO here (sans name).

Reply:
I have converted this into a simple comment.